### PR TITLE
interpret: make some large types not Copy

### PR DESCRIPTION
**Pull request by RalfJung**
Also remove some unused trait impls (mostly HashStable).

This didn't find any unnecessary copies that I managed to avoid, but it might still be better to require explicit clone for these types? Not sure.

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
@Börsig try @rust-timer queue

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 1]
```diff
@@ -459,8 +459,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
                 for i in 0..dest_len {
                     let place = self.mplace_index(&dest, i)?;
-                    let value =
-                        if i == index { *elem } else { self.mplace_index(&input, i)?.into() };
+                    let value = if i == index {
+                        elem.clone()
```
Comment (lqd) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:463]: it's not going to matter much in the grand scheme of things by saving one (1) clone per call to this intrinsic, but this one should be easy to remove by calling `copy_op` in the branches ^^

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:463]: Yeah I considered this and decided, for the reasons you mentioned, that reducing code duplication is more important.

**Comment by oli-obk** [comment]
r=me after a rebase

**Comment by oli-obk** [comment]
weird, the [previous perf run](https://github.com/rust-lang/rust/pull/99309#issuecomment-1186130867) was clean. Inlining threshold noise I guess?

**Comment by oli-obk** [comment]
```
268,061,669  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::run
 -7,078,188  ???:<rustc_middle::ty::subst::SubstFolder as rustc_middle::ty::fold::TypeFolder>::fold_const
 -5,636,093  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::eval_place
  5,453,140  ???:<rustc_middle::ty::instance::Instance>::resolve_opt_const_arg
 -3,829,325  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::eval_fn_call
  3,704,373  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::push_stack_frame
 -3,407,626  ???:<rustc_middle::ty::normalize_erasing_regions::TryNormalizeAfterErasingRegionsFolder as rustc_middle::ty::fold::FallibleTypeFolder>::try_fold_mir_const
 -2,175,874  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::eval_operand
   -367,016  ???:<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::mir_const_to_op
    277,583  ???:rustc_trait_selection::traits::project::normalize::<rustc_middle::ty::sty::TraitRef>
   -277,443  ???:llvm::FPPassManager::runOnFunction(llvm::Function&)
   -262,321  ???:<rustc_target::abi::TyAndLayout<rustc_middle::ty::Ty>>::field::<rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>
```

**Pull request merged**